File: src/utils/fileSystem.ts
import { getStorageAdapter } from '@/adapters/storage';
import { Space, Group, Notebook } from '@/types';
import { parseNoteBlocks, serializeNoteBlocks } from './noteParser';
import { stableIdFromPath } from './stableId';
import { basename, dirname, joinPath, normalizePath } from './path';

function storage() {
  return getStorageAdapter();
}

export async function selectStoragePath(): Promise<string | null> {
  return storage().selectStoragePath();
}

export async function loadSpaces(storagePath: string): Promise<Space[]> {
  const rootPath = normalizePath(storagePath);
  const spaces: Space[] = [];
  let entries;
  try {
    entries = await storage().readDir(rootPath);
  } catch {
    return spaces;
  }

  for (const entry of entries) {
    if (entry.isDirectory && entry.name?.endsWith('.tinynotes')) {
      const spacePath = joinPath(rootPath, entry.name);
      const name = entry.name.replace('.tinynotes', '');
      const children = await loadSpaceChildren(spacePath);
      spaces.push({
        id: stableIdFromPath(spacePath),
        name,
        path: spacePath,
        groups: children,
      });
    }
  }

  return spaces;
}

export async function loadSpaceChildren(spacePath: string): Promise<(Group | Notebook)[]> {
  const parentPath = normalizePath(spacePath);
  const children: (Group | Notebook)[] = [];
  let entries;
  try {
    entries = await storage().readDir(parentPath);
  } catch {
    return children;
  }

  for (const entry of entries) {
    if (entry.isDirectory) {
      const groupPath = joinPath(parentPath, entry.name);
      const subChildren = await loadGroupChildren(groupPath);
      const notebookCount = countNotebooks(subChildren);
      children.push({
        id: stableIdFromPath(groupPath),
        name: entry.name,
        path: groupPath,
        children: subChildren,
        notebookCount,
      });
    } else if (entry.isFile && entry.name?.endsWith('.md')) {
      const filePath = joinPath(parentPath, entry.name);
      const notebook = await loadNotebook(filePath);
      if (notebook) {
        children.push(notebook);
      }
    }
  }

  return children;
}

export async function loadGroups(spacePath: string): Promise<Group[]> {
  const parentPath = normalizePath(spacePath);
  const groups: Group[] = [];
  let entries;
  try {
    entries = await storage().readDir(parentPath);
  } catch {
    return groups;
  }

  for (const entry of entries) {
    if (entry.isDirectory) {
      const groupPath = joinPath(parentPath, entry.name);
      const children = await loadGroupChildren(groupPath);
      const notebookCount = countNotebooks(children);
      groups.push({
        id: stableIdFromPath(groupPath),
        name: entry.name,
        path: groupPath,
        children,
        notebookCount,
      });
    }
  }

  return groups;
}

async function loadGroupChildren(groupPath: string): Promise<(Group | Notebook)[]> {
  const parentPath = normalizePath(groupPath);
  const children: (Group | Notebook)[] = [];
  let entries;
  try {
    entries = await storage().readDir(parentPath);
  } catch {
    return children;
  }

  for (const entry of entries) {
    if (entry.isDirectory) {
      const subGroupPath = joinPath(parentPath, entry.name);
      const subChildren = await loadGroupChildren(subGroupPath);
      const notebookCount = countNotebooks(subChildren);
      children.push({
        id: stableIdFromPath(subGroupPath),
        name: entry.name,
        path: subGroupPath,
        children: subChildren,
        notebookCount,
      });
    } else if (entry.isFile && entry.name?.endsWith('.md')) {
      const filePath = joinPath(parentPath, entry.name);
      const notebook = await loadNotebook(filePath);
      if (notebook) {
        children.push(notebook);
      }
    }
  }

  return children;
}

export function countSpaceNotebooks(space: Space): number {
  return countNotebooks(space.groups);
}

function countNotebooks(children: (Group | Notebook)[]): number {
  let count = 0;
  for (const child of children) {
    if ('noteBlocks' in child) {
      count += 1;
    } else {
      count += countNotebooks((child as Group).children);
    }
  }
  return count;
}

export async function loadNotebook(filePath: string): Promise<Notebook | null> {
  const normalizedPath = normalizePath(filePath);
  try {
    const content = await storage().readTextFile(normalizedPath);
    const noteBlocks = parseNoteBlocks(content, normalizedPath);
    const name = basename(normalizedPath).replace('.md', '');
    return {
      id: stableIdFromPath(normalizedPath),
      name,
      path: normalizedPath,
      noteBlocks,
      isSourceMode: false,
    };
  } catch {
    return null;
  }
}

export async function saveNotebook(notebook: Notebook): Promise<void> {
  const content = serializeNoteBlocks(notebook.noteBlocks);
  await storage().writeTextFile(notebook.path, content);
}

export async function createSpace(storagePath: string, name: string): Promise<Space> {
  const dirName = `${name}.tinynotes`;
  const spacePath = joinPath(storagePath, dirName);
  await storage().mkdir(spacePath, true);
  return {
    id: stableIdFromPath(spacePath),
    name,
    path: spacePath,
    groups: [],
  };
}

export async function createGroup(parentPath: string, name: string): Promise<Group> {
  const groupPath = joinPath(parentPath, name);
  await storage().mkdir(groupPath, true);
  return {
    id: stableIdFromPath(groupPath),
    name,
    path: groupPath,
    children: [],
    notebookCount: 0,
  };
}

export async function createNotebook(parentPath: string, name: string): Promise<Notebook> {
  const fileName = name.endsWith('.md') ? name : `${name}.md`;
  const filePath = joinPath(parentPath, fileName);
  const now = new Date().toISOString();
  const initialContent = `---\ntitle: ${name}\ntags: []\ncreatedAt: ${now}\nupdatedAt: ${now}\n---\n\n`;
  await storage().writeTextFile(filePath, initialContent);
  const notebook = await loadNotebook(filePath);
  if (!notebook) {
    throw new Error('Failed to load created notebook');
  }
  return notebook;
}

export async function duplicateNotebook(sourcePath: string): Promise<Notebook> {
  const normalizedSource = normalizePath(sourcePath);
  const parentPath = dirname(normalizedSource);
  const sourceName = basename(normalizedSource).replace(/\.md$/, '');
  const content = await storage().readTextFile(normalizedSource);

  let copyName = `${sourceName} 副本`;
  let copyPath = joinPath(parentPath, `${copyName}.md`);
  let counter = 2;
  while (await storage().exists(copyPath)) {
    copyName = `${sourceName} 副本 ${counter}`;
    copyPath = joinPath(parentPath, `${copyName}.md`);
    counter++;
  }

  await storage().writeTextFile(copyPath, content);
  const notebook = await loadNotebook(copyPath);
  if (!notebook) {
    throw new Error('Failed to load duplicated notebook');
  }
  return notebook;
}

export async function renameSpace(oldPath: string, newName: string): Promise<string> {
  const parentPath = dirname(oldPath);
  const newPath = joinPath(parentPath, `${newName}.tinynotes`);
  await storage().rename(oldPath, newPath);
  return newPath;
}

export async function deleteSpace(spacePath: string): Promise<void> {
  await storage().remove(spacePath, true);
}

export async function deleteGroup(groupPath: string): Promise<void> {
  await storage().remove(groupPath, true);
}

export async function deleteNotebook(filePath: string): Promise<void> {
  await storage().remove(filePath, false);
}

export async function renameGroup(oldPath: string, newName: string): Promise<string> {
  const parentPath = dirname(oldPath);
  const newPath = joinPath(parentPath, newName);
  await storage().rename(oldPath, newPath);
  return newPath;
}

export async function renameNotebook(oldPath: string, newName: string): Promise<string> {
  const parentPath = dirname(oldPath);
  const newFileName = newName.endsWith('.md') ? newName : `${newName}.md`;
  const newPath = joinPath(parentPath, newFileName);
  await storage().rename(oldPath, newPath);
  return newPath;
}

export async function moveItem(oldPath: string, newParentPath: string): Promise<string> {
  const itemName = basename(oldPath);
  const newPath = joinPath(newParentPath, itemName);
  await storage().rename(oldPath, newPath);
  return newPath;
}

export function getDefaultStoragePath(): string {
  return storage().defaultStoragePath;
}

File: mobile/lib/core/file_system.dart
import 'dart:io';

import '../core/note_parser.dart';
import '../core/path_utils.dart';
import '../core/stable_id.dart';
import '../core/types.dart';
import '../core/workspace_config.dart';
import '../l10n/l10n.dart';
import '../storage/local_storage.dart';

int countNotebooks(List<LibraryItem> children) {
  var count = 0;
  for (final child in children) {
    if (child is NotebookItem) {
      count += 1;
    } else if (child is GroupItem) {
      count += countNotebooks(child.group.children);
    }
  }
  return count;
}

class FileSystemService {
  FileSystemService(this.storage);

  final LocalStorage storage;

  Future<List<Space>> loadSpaces(String storagePath) async {
    final rootPath = normalizePath(storagePath);
    final spaces = <Space>[];
    final entries = await storage.readDir(rootPath);

    for (final entry in entries) {
      if (entry is! Directory) continue;
      final name = basename(entry.path);
      if (!isNoteSpaceDirectoryName(name)) continue;

      final spacePath = joinPath(rootPath, name);
      final children = await loadFolderChildren(spacePath);
      spaces.add(
        Space(
          id: stableIdFromPath(spacePath),
          name: name.replaceAll('.tinynotes', ''),
          path: spacePath,
          groups: children,
        ),
      );
    }

    return spaces;
  }

  Future<List<LibraryItem>> loadFolderChildren(String folderPath) async {
    final parentPath = normalizePath(folderPath);
    final children = <LibraryItem>[];
    final entries = await storage.readDir(parentPath);

    for (final entry in entries) {
      if (entry is Directory) {
        final name = basename(entry.path);
        if (name == workspaceConfigDir) continue;
        final groupPath = joinPath(parentPath, name);
        final subChildren = await loadGroupChildren(groupPath);
        children.add(
          GroupItem(
            Group(
              id: stableIdFromPath(groupPath),
              name: name,
              path: groupPath,
              children: subChildren,
              notebookCount: countNotebooks(subChildren),
            ),
          ),
        );
      } else if (entry is File && entry.path.endsWith('.md')) {
        final notebook = await loadNotebook(entry.path);
        if (notebook != null) {
          children.add(NotebookItem(notebook));
        }
      }
    }

    return children;
  }

  Future<List<LibraryItem>> loadGroupChildren(String groupPath) async {
    final parentPath = normalizePath(groupPath);
    final children = <LibraryItem>[];
    final entries = await storage.readDir(parentPath);

    for (final entry in entries) {
      if (entry is Directory) {
        final name = basename(entry.path);
        final subGroupPath = joinPath(parentPath, name);
        final subChildren = await loadGroupChildren(subGroupPath);
        children.add(
          GroupItem(
            Group(
              id: stableIdFromPath(subGroupPath),
              name: name,
              path: subGroupPath,
              children: subChildren,
              notebookCount: countNotebooks(subChildren),
            ),
          ),
        );
      } else if (entry is File && entry.path.endsWith('.md')) {
        final notebook = await loadNotebook(entry.path);
        if (notebook != null) {
          children.add(NotebookItem(notebook));
        }
      }
    }

    return children;
  }

  Future<Notebook?> loadNotebook(String filePath) async {
    final normalizedPath = normalizePath(filePath);
    try {
      final content = await storage.readTextFile(normalizedPath);
      final noteBlocks = parseNoteBlocks(content, notebookPath: normalizedPath);
      final name = basename(normalizedPath).replaceAll('.md', '');
      return Notebook(
        id: stableIdFromPath(normalizedPath),
        name: name,
        path: normalizedPath,
        noteBlocks: noteBlocks,
      );
    } catch (_) {
      return null;
    }
  }

  Future<void> saveNotebook(Notebook notebook) async {
    final content = serializeNoteBlocks(notebook.noteBlocks);
    await storage.writeTextFile(notebook.path, content);
  }

  Future<Space> createSpace(String storagePath, String name) async {
    final spacePath = joinPath(storagePath, '$name.tinynotes');
    if (await storage.exists(spacePath)) {
      throw StateError(appStrings.fill(appStrings.spaceExists, {'name': name}));
    }
    await storage.mkdir(spacePath, recursive: true);
    return Space(
      id: stableIdFromPath(spacePath),
      name: name,
      path: spacePath,
      groups: const [],
    );
  }

  Future<Group> createGroup(String parentPath, String name) async {
    final groupPath = joinPath(parentPath, name);
    if (await storage.exists(groupPath)) {
      throw StateError(
        appStrings.fill(appStrings.folderExists, {'name': name}),
      );
    }
    await storage.mkdir(groupPath, recursive: true);
    return Group(
      id: stableIdFromPath(groupPath),
      name: name,
      path: groupPath,
      children: const [],
      notebookCount: 0,
    );
  }

  Future<Notebook> createNotebook(String parentPath, String name) async {
    final fileName = name.endsWith('.md') ? name : '$name.md';
    final filePath = joinPath(parentPath, fileName);
    if (await storage.exists(filePath)) {
      throw StateError(
        appStrings.fill(appStrings.notebookExists, {'name': name}),
      );
    }
    final now = DateTime.now().toUtc().toIso8601String();
    final initialContent =
        '---\n'
        'title: $name\n'
        'tags: []\n'
        'createdAt: $now\n'
        'updatedAt: $now\n'
        '---\n\n';
    await storage.writeTextFile(filePath, initialContent);
    final notebook = await loadNotebook(filePath);
    if (notebook == null) {
      throw StateError('Failed to load created notebook');
    }
    return notebook;
  }

  Future<String> renameSpace(String oldPath, String newName) async {
    final parentPath = dirname(oldPath);
    final newPath = joinPath(parentPath, '$newName.tinynotes');
    if (normalizePath(oldPath) == normalizePath(newPath)) return newPath;
    if (await storage.exists(newPath)) {
      throw StateError(
        appStrings.fill(appStrings.spaceExists, {'name': newName}),
      );
    }
    await storage.rename(oldPath, newPath);
    return newPath;
  }

  Future<void> deleteSpace(String spacePath) async {
    await storage.remove(spacePath, recursive: true);
  }

  Future<String> renameGroup(String oldPath, String newName) async {
    final parentPath = dirname(oldPath);
    final newPath = joinPath(parentPath, newName);
    if (normalizePath(oldPath) == normalizePath(newPath)) return newPath;
    if (await storage.exists(newPath)) {
      throw StateError(
        appStrings.fill(appStrings.folderExists, {'name': newName}),
      );
    }
    await storage.rename(oldPath, newPath);
    return newPath;
  }

  Future<String> renameNotebook(String oldPath, String newName) async {
    final parentPath = dirname(oldPath);
    final newFileName = newName.endsWith('.md') ? newName : '$newName.md';
    final newPath = joinPath(parentPath, newFileName);
    if (normalizePath(oldPath) == normalizePath(newPath)) return newPath;
    if (await storage.exists(newPath)) {
      throw StateError(
        appStrings.fill(appStrings.notebookExists, {'name': newName}),
      );
    }
    await storage.rename(oldPath, newPath);
    return newPath;
  }

  Future<void> deleteGroup(String groupPath) async {
    await storage.remove(groupPath, recursive: true);
  }

  Future<void> deleteNotebook(String filePath) async {
    await storage.remove(filePath, recursive: false);
  }

  Future<void> ensureSampleLibrary(String storagePath) async {
    final spaces = await loadSpaces(storagePath);
    if (spaces.isNotEmpty) return;

    final s = appStrings;
    final space = await createSpace(storagePath, s.sampleSpaceName);
    final group = await createGroup(space.path, s.sampleGroupName);
    final notebook = await createNotebook(group.path, s.sampleNotebookName);
    final now = DateTime.now().toUtc().toIso8601String();

    await saveNotebook(
      Notebook(
        id: notebook.id,
        name: notebook.name,
        path: notebook.path,
        noteBlocks: [
          NoteBlock(
            id:
                notebook.noteBlocks.isNotEmpty
                    ? notebook.noteBlocks.first.id
                    : 'sample-1',
            title: s.samplePortTitle,
            content: 'lsof -i :8080',
            contentType: ContentType.bash,
            tags: const ['shell'],
            createdAt: now,
            updatedAt: now,
          ),
          NoteBlock(
            id: 'sample-2',
            title: s.sampleGitStatusTitle,
            content: 'git status -sb',
            contentType: ContentType.bash,
            tags: const ['git'],
            createdAt: now,
            updatedAt: now,
          ),
        ],
      ),
    );
  }
}
